chapter_13_database_layer_validation.asciidoc

```diff
@@ -81,155 +81,270 @@ This test is interesting
+(and change the test name):
+
+
+[role="sourcecode"]
+.src/lists/tests/test_models.py (ch13l002)
+====
+[source,python]
+----
+    def test_cannot_save_null_list_items(self):
         mylist = List.objects.create()
         item = Item(list=mylist, text=None)
-        item.save()
-////
+        with self.assertRaises(IntegrityError):
+            item.save()
+----
+====
+
+You'll see that _this_ test now passes.
+
+----
+Ran 10 tests in 0.030s
+
+OK
+----
+
+
+==== Testing Django Model Validation:
+
+That's all vaguely interesting, but it's not actually what we set out to do.
+How do we make sure that the "validation-related" constraint is being enforced?
+The answer is that, while `IntegrityError` comes from the database,
+Django uses `ValidationError` to signal errors that come from its own validation.
+
+Let's write a second test that checks on that:
+
+
+[role="sourcecode"]
+.src/lists/tests/test_models.py (ch13l003)
+====
+[source,python]
+----
+from django.core.exceptions import ValidationError
+from django.db.utils import IntegrityError
+[...]
 
+class ListAndItemModelsTest(TestCase):
+    def test_saving_and_retrieving_items(self):
+        [...]
+
+    def test_cannot_save_null_list_items(self):
+        mylist = List.objects.create()
+        item = Item(list=mylist, text=None)
+        with self.assertRaises(IntegrityError):
+            item.save()
+
+    def test_cannot_save_empty_list_items(self):
+        mylist = List.objects.create()
+        item = Item(list=mylist, text="")  # <1>
+        with self.assertRaises(ValidationError):  # <2>
+            item.save()
+----
+====
+
+<1> This time we pass `text=""`
+<2> And we're expecting a `ValidationError` instead of an `IntegrityError`:
+
+
+
+==== A Django Quirk: Model Save Doesn't Run Validation
+
+We can try running _this_ new unit test,
+and we'll see its expected failure...
+
+----
+    with self.assertRaises(ValidationError):
+AssertionError: ValidationError not raised
+----
+
+Wait a minute!  We expected this to _pass_ actually!
+We just got through learning that Django should be enforcing the `blank=False`
+constraint by default.  Why doesn't this work?
+
+((("model-layer validation", "running full validation")))
+We've discovered one of Django's little quirks.
+For
+https://groups.google.com/g/django-developers/c/uIhzSwWHj4c[slightly
+counterintuitive historical reasons],
+Django models don't run full validation on save.
 
 ((("full_clean method")))
 Django does have a method to manually run full validation, however,
 called `full_clean` (more info in
 https://docs.djangoproject.com/en/4.2/ref/models/instances/#django.db.models.Model.full_clean[the docs]).
-Let's hack it in to see it work:
+Let's add it in to see it work:
 
 
 [role="sourcecode"]
-.src/lists/tests/test_models.py
+.src/lists/tests/test_models.py (ch13l004)
 ====
 [source,python]
 ----
     with self.assertRaises(ValidationError):
         item.full_clean()
-        item.save()
```

Comment:
are we removing the `save()` and changing it to a `full_clean()` or are we just inserting the `full_clean` in front of `save()`?

Reply:
changing it.  i'll add a note!